src/protocol/libp2p/kademlia/mod.rs

```diff
@@ -642,50 +642,71 @@ impl Kademlia {
-            {
-                Err(_) => {
-                    tracing::trace!(target: LOG_TARGET, ?query, ?peer, "dial peer");
-
-                    match self.service.dial(&peer) {
-                        Ok(_) => match self.pending_dials.entry(peer) {
-                            Entry::Occupied(entry) => {
-                                entry.into_mut().push(PeerAction::SendFindNode(query));
-                            }
-                            Entry::Vacant(entry) => {
-                                entry.insert(vec![PeerAction::SendFindNode(query)]);
-                            }
-                        },
-                        Err(error) => {
-                            tracing::trace!(target: LOG_TARGET, ?query, ?peer, ?error, "failed to dial peer");
-                            self.engine.register_response_failure(query, peer);
-                        }
+    /// Open a substream with a peer or dial the peer.
+    fn open_substream_or_dial(
+        &mut self,
+        peer: PeerId,
+        action: PeerAction,
+        query: Option<QueryId>,
+    ) -> Result<(), Error> {
+        match self.service.open_substream(peer) {
+            Ok(substream_id) => {
+                self.pending_substreams.insert(substream_id, peer);
+                self.peers.entry(peer).or_default().pending_actions.insert(substream_id, action);
+
+                return Ok(());
+            }
+            Err(err) => {
+                tracing::trace!(target: LOG_TARGET, ?query, ?peer, ?err, "Failed to open substream. Dialing peer");
+
+                match self.service.dial(&peer) {
+                    Ok(()) => {
+                        self.pending_dials.entry(peer).or_default().push(action);
+                        return Ok(());
                     }
 
-                    Ok(())
-                }
-                Ok(substream_id) => {
-                    tracing::trace!(
-                        target: LOG_TARGET,
-                        ?query,
-                        ?peer,
-                        ?substream_id,
-                        "open outbound substream for peer"
-                    );
+                    // Already connected is a recoverable error.
+                    Err(Error::AlreadyConnected) => (),
 
-                    self.pending_substreams.insert(substream_id, peer);
-                    self.peers
-                        .entry(peer)
-                        .or_default()
-                        .pending_actions
-                        .insert(substream_id, PeerAction::SendFindNode(query));
+                    Err(error) => {
+                        tracing::trace!(target: LOG_TARGET, ?query, ?peer, ?error, "Failed to dial peer");
+                        return Err(error);
+                    }
+                };
 
-                    Ok(())
+                // Dial returned `Error::AlreadyConnected`, retry opening the substream.
```

Comment:
nit: Does this block indents badly if put into `Err(Error::AlreadyConnected)` arm? This way we won't need returns and it would be slightly easier to follow the logic. Feel free to ignore, it's likely a matter of personal taste.

Reply:
Yep that makes things clearer! Thanks for the suggestion 🙏